server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,260 @@
+            readyToPublish = false;
+        }
+    }
+
+    private void updateOnClusterStateChange(ClusterChangedEvent event) {
+        final boolean wasMaster = event.previousState().nodes().isLocalNodeElectedMaster();
+        isMaster = event.localNodeMaster();
+        if (isMaster && wasMaster == false) {
+            readyToPublish = true;
+        } else if (isMaster == false) {
+            readyToPublish = false;
+        }
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            if (readyToPublish) {
+                resetHealthMetadata("health-metadata-update-master-election");
+                readyToPublish = false;
+            }
+        }
+    }
+
+    private void updateOnSettingsUpdated(String setting, String value) {
+        // We do not use the cluster state to check if this is the master node because the cluster state might not have been initialized
+        if (isMaster && enabled) {
+            ClusterState clusterState = clusterService.state();
+            if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+                var task = new UpdateHealthMetadata(setting, value);
+                var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+                clusterService.submitStateUpdateTask("health-metadata-update", task, config, executor);
+            }
+        }
+    }
+
+    private void resetHealthMetadata(String source) {
+        var task = new InsertHealthMetadata(settings);
+        var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+        clusterService.submitStateUpdateTask(source, task, config, executor);
+    }
+
+    public static List<NamedXContentRegistry.Entry> getNamedXContentParsers() {
+        return List.of(
+            new NamedXContentRegistry.Entry(Metadata.Custom.class, new ParseField(HealthMetadata.TYPE), HealthMetadata::fromXContent)
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    /**
+     * A base class for health metadata cluster state update tasks.
+     */
+    abstract static class UpsertHealthMetadataTask implements ClusterStateTaskListener {
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "never called";
+        }
+
+        @Override
+        public void onFailure(@Nullable Exception e) {
+            logger.error("failure during health metadata update", e);
+        }
+
+        abstract HealthMetadata execute(HealthMetadata initialHealthMetadata);
```

Comment:
Should the param be marked with @Nullable?

Reply:
Or maybe avoid the ignored parameter in the `Insert...` implementation by having the `ClusterState currentState` as parameter? (each implementation can then look for the `health` custom and add it/modify it)

Maybe even return `ClusterState` ?

Reply:
Ah, even better!!! Thank you for the suggestion!